### ISSUE TITLE
Avoid publishing new version if nothing has changed

### ISSUE THREAD

**Issue by szczepiq** [mockito/mockito-release-tools#84]
## Problem

Publishing new version when it is binary equal to the previous version is confusing for the clients. At the moment we don't check if new version is different than the current release. If someone makes a change in .gitignore it can trigger build and potentially a publication.

## Suggested implementation

Only a suggested plan, please execute as you see fit!

### Robustness of the algorithm

The comparison algorithm is _not_ on the critical path and we are allowed to be pragmatic in coming up the implementation and we can iterate on the solution to make it better. Here are the effect of false positive / false negative signal from the algorithm:
 - false negative, publications incorrectly considered non-equal: extra version is released that does not have changes on top of previous version. It is minor problem for projects that leverage the "notable releases" feature. For other projects it should not be a huge deal because versions are shipped frequently in continuous delivery.
 - false positive, publications incorrectly considered equal: release is not triggered and needs to be triggered semi-manually. Not a critical problem.

### Gradle plugins / tasks

1. We have a "releaseNeeded" task that determines whether the release should be triggered. This task needs to depend on some other task(s) that pull down and compare publications.
2. We could create "PublicationsComparatorPlugin" that applies "BaseJavaLibraryPlugin" and "VersioningPlugin". New plugin would add task that that compares publications. New plugin would be applied by "JavaLibraryPlugin".
3. releaseNeeded would depend on all tasks added at step 2) from projects that use PublicationsComparatorPlugin.

### Implementation

1. It is problematic to use Gradle's configurations to download previous versions of jars / poms for comparisons because Gradle automatically replaces dependencies with local project dependency with the version that is currently being built. We could reach out to Gradle folks on the forum and ask how to do it. If we cannot get it to work, we need to hit the url directly and download file without using Gradle's dependency resolution.
2. The task should be incremental, e.g. when nothing changes and build dir is intact, we don't download jars again.
3. Current publication model is that we we publish all jars from the project during the release. We want to stick with this model of releasing because it is simple and powerful - the project is a unit of deployment/release and all jars go out with the same version. If only one jar has changed, we still publish all jars with new version.
4. Stuff to consider when comparing publications:
 - pom.xml content
 - jar content / binaries (for that, perhaps it's simpler to just compare the sources jar, it will cover compiled classes and javadocs at the same time)
 - MANIFEST.MF from jars
5. There is already some code to be reused, legacy of Mockito automation :) See "org.mockito.release.internal.comparison" package.

**Comment by wwilk** [comment]
@szczepiq 
1. Should we start with configuration for Bintray and later add one for other repositories? I thought about closure that you would pass in configuration of task, that would return artifact URL based on a few parameters - version, group, name, suffix (eg. .pom or -sources.jar). 

2. Currently BaseJavaLibraryPlugin doesn't depend on VersioningPlugin, the problem with that is that we need to access version and previousVersion to compare artifacts. If we apply VersioningPlugin there another issue comes up - version.properties is created in subprojects. How should we handle it? Since we want to have the same version for all projects we can skip creation of version.properties (and also bumping of version since it needs to be bumped only once) in subprojects. What do you think about this approach? Can we even apply VersioningPlugin in JavaLibraryPlugin?

**Comment by szczepiq** [comment]
>Should we start with configuration for Bintray and later add one for other repositories?

- it's best if feature works out of the box for the core use case (mockito, our example projects)
- for other types of projects, let's make things configurable so that the feature is useful. We can get there iteratively, no necessarily we need to bake one PR to rule them all :)

>Currently BaseJavaLibraryPlugin doesn't depend on VersioningPlugin, the problem with that is that we need to access version and previousVersion to compare artifacts. If we apply VersioningPlugin there another issue comes up - version.properties is created in subprojects. How should we handle it?

Great catch, I missed that in the design spec. How about we apply similar patter to VersioningPlugin as we do for the ReleaseConfigurationPlugin? The latter plugin is only ever added to the root project.

@wwilk, also see my comments to your PR.
